cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AppLifecycleHttpHandler.java

```diff
@@ -252,7 +259,15 @@ public void deleteApp(HttpRequest request, HttpResponder responder,
                         @PathParam("namespace-id") String namespaceId,
                         @PathParam("app-id") final String appId) {
     try {
-      Id.Program id = Id.Program.from(namespaceId, appId, "");
+      Id.Application id = Id.Application.from(namespaceId, appId);
+
+      // Deletion of a particular application is not allowed if that application is used by an adapter
+      if (adapterService.getAdapterTypeInfo(appId) != null) {
```

Comment:
it's looking more and more like we need some property to distinguish system apps vs. user apps, but lets handle it outside of this PR.